### PR TITLE
Issue 43442: samples: not able to delete a Material not in a sample type

### PR DESCRIPTION
**Pull request by labkey-kevink**
#### [Issue 43442](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=43442): samples: not able to delete a Material not in a sample type
* SampleTypeUpdateServiceDI needs to handle read and delete without a ExpSampleType
* Add regression test to verify read, insert, update, delete permissions

#### [Issue 43407](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=43407): Required boolean fields cannot be set to false in the sample derivation view
* The action uses custom form binding of the posted parameters and didn't handle the special `@` field prefix for checkbox input parameters.

#### [Issue 43439](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=43439): LK Server Allows Insert of Samples Outside a Specific Sample Type
* Remove the option to insert into exp.Materials
* Validate a SampleType was chosen

#### Bonus Round
* Add `SqlFragment.appendIf` for all those conditional sql woes
* Remove obsolete `~checkboxes` parameter handling
* Clear selection after deleting samples

### CODE REVIEW

File: api/src/org/labkey/api/data/SQLFragment.java [hunk 1]
```diff
@@ -566,6 +567,19 @@ public CharSequence subSequence(int start, int end)
...
     }
 
+    public SQLFragment appendIf(boolean test, CharSequence s)
+    {
+        if (test)
+            this.append(s);
+        return this;
+    }
+
+    public SQLFragment appendIf(boolean test, Consumer<SQLFragment> block)
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/data/SQLFragment.java]: The naming of this one is odd??? I'd have expected appendIf(test, Supplier<SQLFragment>), I'm not sure what I'd call this, something like evalIf()?

File: experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java [hunk 2]
```diff
@@ -799,11 +801,19 @@ public QueryUpdateService getUpdateService()
     @Override
     public boolean hasPermission(@NotNull UserPrincipal user, @NotNull Class<? extends Permission> perm)
     {
-        if (_ss != null)
+        if (_ss == null)
+        {
+            // Allow read and delete for exp.Materials.
+            // Don't allow insert/update on exp.Materials without a sample type.
+            if (perm.isAssignableFrom(DeletePermission.class) || perm.isAssignableFrom(ReadPermission.class))
```
Comment (labkey-matthewb) [inline experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java]: I think == is appopriate here

File: experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java [hunk 3]
```diff
@@ -386,7 +400,6 @@ protected Domain getDomain()
     public List<Map<String, Object>> deleteRows(User user, Container container, List<Map<String, Object>> keys, @Nullable Map<Enum, Object> configParameters, @Nullable Map<String, Object> extraScriptContext)
```
Comment (labkey-matthewb) [inline experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java:400]: When _sampleType==null, are we enforcing that the deleted rows _do not_ belong to a sampletype?

File: experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java [hunk 4]
```diff
@@ -139,12 +139,14 @@ User getUser()
     protected DataIteratorBuilder preTriggerDataIterator(DataIteratorBuilder in, DataIteratorContext context)
     {
         // MOVE PrepareDataIteratorBuilder into this file
+        assert _sampleType != null : "SampleType required for insert/update, but not required for read/delete";
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java:142]: preTriggerDataIterator is called only for insert/update?

File: experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java [hunk 5]
```diff
@@ -233,12 +240,15 @@ public int mergeRows(User user, Container container, DataIteratorBuilder rows, B
...
-        SQLFragment sql = new SQLFragment()
-                .append("SELECT t.*, d.RowId, d.Name, d.Container, d.Description, d.CreatedBy, d.Created, d.ModifiedBy, d.Modified")
+        SQLFragment sql = new SQLFragment("SELECT ")
+                .appendIf(t != null, "t.*,")
+                .append(" d.RowId, d.Name, d.Container, d.Description, d.CreatedBy, d.Created, d.ModifiedBy, d.Modified")
                 .append(" FROM ").append(d, "d")
-                .append(" LEFT OUTER JOIN ").append(t, "t")
+                .applyIf(t != null, s ->
+                    s.append("LEFT OUTER JOIN ").append(t, "t")
+                )
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java:251]: The " ON" clause that follows should also be included in the `applyIf` block.

**Review by labkey-susanh — APPROVED** [review]
Looks like good changes for the "no sample type" case.  I do still wonder when this gets used.

One minor change needed for the SQL construction.

**Comment by labkey-kevink** [comment]
Moved to #2423 for 21.7 release

**Pull request closed without merging**